cache/blobs_linux.go

```diff
@@ -0,0 +1,394 @@
+			// File doesn't exist in the base layer. Thus this is added.
+			kind = fs.ChangeKindAdd
+		} else if err != nil {
+			return err
+		}
+
+		if !skipRecord {
+			if err := changeFn(kind, path, f, nil); err != nil {
+				return err
+			}
+		}
+
+		if f != nil {
+			if isOpaque, err := checkOpaque(upperdir, path, base, f); err != nil {
+				return err
+			} else if isOpaque {
+				// This is an opaque directory. Start a new walking differ to get adds/deletes of
+				// this directory. We use "upperdirView" directory which doesn't contain whiteouts.
+				if err := fs.Changes(ctx, filepath.Join(base, path), filepath.Join(upperdirView, path),
+					func(k fs.ChangeKind, p string, f os.FileInfo, err error) error {
+						return changeFn(k, filepath.Join(path, p), f, err) // rebase path to be based on the opaque dir
+					},
+				); err != nil {
+					return err
+				}
+				return filepath.SkipDir // We completed this directory. Do not walk files under this directory anymore.
+			}
+		}
+		return nil
+	})
+}
+
+// checkDelete checks if the specified file is a whiteout
+func checkDelete(upperdir string, path string, base string, f os.FileInfo) (delete, skip bool, _ error) {
+	if f.Mode()&os.ModeCharDevice != 0 {
+		if _, ok := f.Sys().(*syscall.Stat_t); ok {
+			maj, min, err := devices.DeviceInfo(f)
+			if err != nil {
+				return false, false, errors.Wrapf(err, "failed to get device info")
+			}
+			if maj == 0 && min == 0 {
+				// This file is a whiteout (char 0/0) that indicates this is deleted from the base
+				if _, err := os.Lstat(filepath.Join(base, path)); err != nil {
+					if !os.IsNotExist(err) {
+						return false, false, errors.Wrapf(err, "failed to lstat")
+					}
+					// This file doesn't exist even in the base dir.
+					// We don't need whiteout. Just skip this file.
+					return false, true, nil
+				}
+				return true, false, nil
+			}
+		}
+	}
+	return false, false, nil
+}
+
+// checkDelete checks if the specified file is an opaque directory
+func checkOpaque(upperdir string, path string, base string, f os.FileInfo) (isOpaque bool, _ error) {
+	if f.IsDir() {
+		for _, oKey := range []string{"trusted.overlay.opaque", "user.overlay.opaque"} {
+			opaque, err := sysx.LGetxattr(filepath.Join(upperdir, path), oKey)
+			if err != nil && err != unix.ENODATA {
+				return false, errors.Wrapf(err, "failed to retrieve %s attr", oKey)
+			} else if len(opaque) == 1 && opaque[0] == 'y' {
+				// This is an opaque whiteout directory.
+				if _, err := os.Lstat(filepath.Join(base, path)); err != nil {
+					if !os.IsNotExist(err) {
+						return false, errors.Wrapf(err, "failed to lstat")
+					}
+					// This file doesn't exist even in the base dir. We don't need treat this as an opaque.
+					return false, nil
+				}
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+// sameDir performs continity-compatible comparison of directories.
+// https://github.com/containerd/continuity/blob/v0.1.0/fs/path.go#L91-L133
```

Comment:
Can the code that is already in continuity be deduplicated later?

Reply:
I'll open PRs in continuity to make them public functions.